yt_dlp/extractor/bilibili.py

```diff
@@ -879,12 +899,25 @@ def _real_extract(self, url):
                 'Extracting episode', query={'fnval': 12240, 'ep_id': episode_id},
                 headers=headers))
 
+        geo_blocked = traverse_obj(play_info, ('raw', 'data', 'plugins', lambda _, v: v['name'] == 'AreaLimitPanel', 'config', 'is_block', {bool}, any))
         premium_only = play_info.get('code') == -10403
-        play_info = traverse_obj(play_info, ('result', 'video_info', {dict})) or {}
 
-        formats = self.extract_formats(play_info)
-        if not formats and (premium_only or '成为大会员抢先看' in webpage or '开通大会员观看' in webpage):
-            self.raise_login_required('This video is for premium members only')
+        video_info = traverse_obj(play_info, (('result', ('raw', 'data')), 'video_info', {dict}, any)) or {}
+        formats = self.extract_formats(video_info)
+
+        if not formats:
+            if geo_blocked:
+                self.raise_geo_restricted()
+            elif premium_only or '成为大会员抢先看' in webpage or '开通大会员观看' in webpage:
+                self.raise_login_required('This video is for premium members only')
+
+        if traverse_obj(play_info, (
+            'result', 'play_check', 'play_detail', {lambda x: x and x == 'PLAY_PREVIEW'}),  # vs 'PLAY_WHOLE'
+            ('raw', 'data', 'play_video_type', {lambda x: x and x == 'preview'}),  # vs 'whole'
```

Comment:
I *think* the `x and` is unnecessary? But lemme know if it's not
```suggestion
            'result', 'play_check', 'play_detail', {lambda x: x == 'PLAY_PREVIEW'}),  # vs 'PLAY_WHOLE'
            ('raw', 'data', 'play_video_type', {lambda x: x == 'preview'}),  # vs 'whole'
```

Reply:
`x and x == ...` is necessary because otherwise when `x == None`, `None == 'preview'` always gives False.
Perhaps we can do `x == 'whole'` and `x == 'PLAY_WHOLE'`, but I am aligned to just allow this to be a little longer
```
(Pdb) traverse_obj(play_info, ('result', 'play_check', 'play_detail', {lambda x: x and x == 'PLAY_PREVIEW'}))
(Pdb) traverse_obj(play_info, ('result', 'play_check', 'play_detail', {lambda x: x == 'PLAY_PREVIEW'}))       
False
(Pdb) traverse_obj(play_info, ('raw', 'data', 'play_video_type', {lambda x: x and x == 'preview'}))  
True
(Pdb) traverse_obj(play_info, ('result', 'play_check', 'play_detail', {lambda x: x == 'PLAY_PREVIEW'}),('raw', 'data', 'play_video_type', {lambda x:x == 'preview'}))
False
(Pdb)
```

Reply:
Ah right nevermind